src/lib/SaferNames/Type.hs

```diff
@@ -0,0 +1,968 @@
+        _ -> throw TypeErr $
+              "Only single-member ADTs and record types can be projected. Got " <> pprint ty <> "   " <> pprint v
+
+instance CheckableB Binder where
+  checkB (b:>ty) cont = do
+    ty' <- checkTypeE TyKind ty
+    withFreshBinder b ty' MiscBound \b' ->
+      extendRenamer (b@>binderName b') $
+        cont b'
+
+instance HasType Expr where
+  getTypeE expr = case expr of
+    App f x -> do
+      fTy <- getTypeE f
+      checkApp fTy x
+    Atom x   -> getTypeE x
+    Op   op  -> typeCheckPrimOp op
+    Hof  hof -> typeCheckPrimHof hof
+    Case e alts resultTy -> checkCase e alts resultTy
+
+instance HasType Block where
+  getTypeE (Block ty decls expr) = do
+    tyReq <- substM ty
+    checkB decls \decls' -> do
+      tyReq' <- injectM decls' tyReq
+      expr |: tyReq'
+    return tyReq
+
+instance CheckableB Decl where
+  checkB (Let ann (b:>ty) expr) cont = do
+    ty' <- checkTypeE TyKind ty
+    expr' <- checkTypeE ty' expr
+    withFreshBinder b ty' (LetBound ann expr') \b' ->
+      extendRenamer (b @> binderName b') $
+        cont $ Let ann b' expr'
+
+instance CheckableB b => CheckableB (Nest b) where
+  checkB nest cont = case nest of
+    Empty -> cont Empty
+    Nest b rest -> checkB b \b' -> checkB rest \rest' -> cont $ Nest b' rest'
+
+typeCheckPrimTC :: Typer m => PrimTC (Atom i) -> m i o (Type o)
+typeCheckPrimTC tc = case tc of
+  BaseType _       -> return TyKind
+  IntRange a b     -> a|:IdxRepTy >> b|:IdxRepTy >> return TyKind
+  IndexRange t a b -> do
+    t' <- checkTypeE TyKind t
+    mapM_ (|:t') a >> mapM_ (|:t') b >> return TyKind
+  IndexSlice n l   -> n|:TyKind >> l|:TyKind >> return TyKind
+  ProdType tys     -> mapM_ (|:TyKind) tys >> return TyKind
+  SumType  cs      -> mapM_ (|:TyKind) cs  >> return TyKind
+  RefType r a      -> mapM_ (|:TyKind) r >> a|:TyKind >> return TyKind
+  TypeKind         -> return TyKind
+  EffectRowKind    -> return TyKind
+  LabeledRowKindTC -> return TyKind
+  ParIndexRange t gtid nthr -> gtid|:IdxRepTy >> nthr|:IdxRepTy >> t|:TyKind >> return TyKind
+
+typeCheckPrimCon :: Typer m => PrimCon (Atom i) -> m i o (Type o)
+typeCheckPrimCon con = case con of
+  Lit l -> return $ BaseTy $ litType l
+  ProdCon xs -> ProdTy <$> mapM getTypeE xs
+  SumAsProd ty tag _ -> tag |:TagRepTy >> substM ty  -- TODO: check!
+  ClassDictHole _ ty  -> ty |:TyKind   >> substM ty
+  IntRangeVal     l h i -> i|:IdxRepTy >> substM (TC $ IntRange     l h)
+  IndexRangeVal t l h i -> i|:IdxRepTy >> substM (TC $ IndexRange t l h)
+  IndexSliceVal _ _ _ -> error "not implemented"
+  BaseTypeRef p -> do
+    (PtrTy (_, b)) <- getTypeE p
+    return $ RawRefTy $ BaseTy b
+  TabRef tabTy -> do
+    TabTy b (RawRefTy a) <- getTypeE tabTy
+    return $ RawRefTy $ TabTy b a
+  ConRef conRef -> case conRef of
+    ProdCon xs -> RawRefTy <$> (ProdTy <$> mapM typeCheckRef xs)
+    IntRangeVal     l h i -> do
+      l' <- substM l
```

Comment:
Can you please remind me why does type checking now need a substitution? I briefly recall us discussion that but I already forgot 😕

Reply:
Unless you have the `Distinct n` constraint, shadowing is possible. So you can end up with expressions like this:
```
\(x:a). \(a:Type). x
```
The type of `x` should be the outer `a`, not the inner one.

Previously, we sort-of-mostly tried to enforce the no-shadow invariant everywhere, so you had to deshadow when looking up a substitution, for example. We also previously had types cached alongside variable occurrences, so the above example would have been impossible.